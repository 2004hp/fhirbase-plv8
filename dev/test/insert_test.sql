--db:fhirb -e
--{{{
\set obs `cat test/fixtures/observation-example-f001-glucose.json`
\set uuid '550e8400-e29b-41d4-a716-446655440000'
\set tags '[{"scheme": "fhir.com", "term": "term", "label": "label"},{"scheme": "fhir.com1", "term": "term1", "label": "label1"}]'
\set new_tags '[{"scheme": "fhir.com", "term": "term", "label": "label"}, {"scheme": "fhir.com2", "term": "term2", "label": "label2"}]'

BEGIN;

SELECT assert(count(*) = 0, 'empty observation') FROM observation where logical_id = :'uuid';
SELECT assert(count(*) = 0, 'empty observation_history') FROM observation_history where logical_id = :'uuid';
SELECT assert(count(*) = 0, 'empty observation_tag') FROM observation_tag where resource_id = :'uuid';
SELECT assert(count(*) = 0, 'empty observation_tag_history') FROM observation_tag_history where resource_id = :'uuid';

SELECT insert_resource(:'uuid'::uuid, :'obs'::jsonb, :'tags'::jsonb);

SELECT assert(count(*) = 1, 'insert: not empty observation') FROM observation where logical_id = :'uuid';
SELECT assert(count(*) = 0, 'insert: empty history') FROM observation_history where logical_id = :'uuid';
SELECT assert(count(*) = 2, 'insert: not empty tags') FROM observation_tag where resource_id = :'uuid';
SELECT assert(count(*) = 0, 'insert: empty tags history') FROM observation_tag_history where resource_id = :'uuid';

SELECT update_resource(:'uuid'::uuid, :'obs'::jsonb, :'new_tags'::jsonb);

SELECT assert(count(*)=1,'update: not empty observation') FROM observation where logical_id = :'uuid';
SELECT assert(count(*)=1,'update: not empty history') FROM observation_history where logical_id = :'uuid';
SELECT assert(count(*)=3,'update: 3 merged tags') FROM observation_tag where resource_id = :'uuid';
SELECT assert(count(*)=2,'update: 2 tags in history') FROM observation_tag_history where resource_id = :'uuid';

SELECT delete_resource(:'uuid'::uuid, 'observation');

SELECT assert(count(*)=0,'delete: empty res') FROM observation where logical_id = :'uuid';
SELECT assert(count(*)=2,'delete: 2 in history') FROM observation_history where logical_id = :'uuid';
SELECT assert(count(*)=0,'delete: 0 in tags') FROM observation_tag where resource_id = :'uuid';
SELECT assert(count(*)=5,'delete: 5 in history tags') FROM observation_tag_history where resource_id = :'uuid';

ROLLBACK;
--}}}
