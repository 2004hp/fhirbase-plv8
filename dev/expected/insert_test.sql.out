BEGIN;
BEGIN
SELECT assert(count(*) = 0, 'empty observation') FROM observation where logical_id = '550e8400-e29b-41d4-a716-446655440000';
        assert        
----------------------
 OK empty observation
(1 row)

SELECT assert(count(*) = 0, 'empty observation_history') FROM observation_history where logical_id = '550e8400-e29b-41d4-a716-446655440000';
            assert            
------------------------------
 OK empty observation_history
(1 row)

SELECT assert(count(*) = 0, 'empty observation_tag') FROM observation_tag where resource_id = '550e8400-e29b-41d4-a716-446655440000';
          assert          
--------------------------
 OK empty observation_tag
(1 row)

SELECT assert(count(*) = 0, 'empty observation_tag_history') FROM observation_tag_history where resource_id = '550e8400-e29b-41d4-a716-446655440000';
              assert              
----------------------------------
 OK empty observation_tag_history
(1 row)

SELECT insert_resource('550e8400-e29b-41d4-a716-446655440000'::uuid,  E'{"resourceType":"Observation","text":{"status":"generated","div":"<div>\\n      <p>\\n        <b>Generated Narrative</b>\\n      </p>\\n      <p>\\n        <b>name</b>: \\n        <span title=\\"Codes: {http://loinc.org 2339-0}\\">Glucose [Mass/volume] in Blood</span>\\n      </p>\\n      <p>\\n        <b>value[x]</b>: 6.3 mmol/l\\n      </p>\\n      <p>\\n        <b>interpretation</b>: \\n        <span title=\\"Codes: {http://hl7.org/fhir/v2/0078 A}\\">abnormal</span>\\n      </p>\\n      <p>\\n        <b>applies[x]</b>: 2-Apr 2013 9:30 --&gt; 5-Apr 2013 9:30\\n      </p>\\n      <p>\\n        <b>issued</b>: 3-Apr 2013 15:30\\n      </p>\\n      <p>\\n        <b>status</b>: final_\\n      </p>\\n      <p>\\n        <b>reliability</b>: ok\\n      </p>\\n      <p>\\n        <b>bodySite</b>: \\n        <span title=\\"Codes: {http://snomed.info/sct 308046002}\\">Superficial forearm vein</span>\\n      </p>\\n      <p>\\n        <b>method</b>: \\n        <span title=\\"Codes: {http://snomed.info/sct 120220003}\\">Injection to forearm</span>\\n      </p>\\n      <p>\\n        <b>identifier</b>: 6323 (official)\\n      </p>\\n      <p>\\n        <b>subject</b>: P. van de Heuvel\\n      </p>\\n      <p>\\n        <b>performer</b>: A. Langeveld\\n      </p>\\n      <h3>ReferenceRanges</h3>\\n      <table class=\\"grid\\">\\n        <tr>\\n          <td>\\n            <b>Low</b>\\n          </td>\\n          <td>\\n            <b>High</b>\\n          </td>\\n          <td>\\n            <b>Meaning</b>\\n          </td>\\n          <td>\\n            <b>Age</b>\\n          </td>\\n        </tr>\\n        <tr>\\n          <td>3.1 mmol/l</td>\\n          <td>6.2 mmol/l</td>\\n          <td> </td>\\n          <td> </td>\\n        </tr>\\n      </table>\\n    </div>"},"name":{"coding":[{"system":"http://loinc.org","code":"2339-0","display":"Glucose [Mass/volume] in Blood"}]},"valueQuantity":{"value":6.3,"units":"mmol/l","system":"http://unitsofmeasure.org","code":"mmol/l"},"interpretation":{"coding":[{"system":"http://hl7.org/fhir/v2/0078","code":"A","display":"abnormal"}]},"appliesPeriod":{"start":"2013-04-02T09:30:10+01:00","end":"2013-04-05T09:30:10+01:00"},"issued":"2013-04-03T15:30:10+01:00","status":"final","reliability":"ok","bodySite":{"coding":[{"system":"http://snomed.info/sct","code":"308046002","display":"Superficial forearm vein"}]},"method":{"coding":[{"system":"http://snomed.info/sct","code":"120220003","display":"Injection to forearm"}]},"identifier":{"use":"official","system":"http://www.bmc.nl/zorgportal/identifiers/observations","value":"6323"},"subject":{"reference":"Patient/f001","display":"P. van de Heuvel"},"performer":[{"reference":"Practitioner/f005","display":"A. Langeveld"}],"referenceRange":[{"low":{"value":3.1,"units":"mmol/l","system":"http://unitsofmeasure.org","code":"mmol/l"},"high":{"value":6.2,"units":"mmol/l","system":"http://unitsofmeasure.org","code":"mmol/l"}}]}'::jsonb, '[{"scheme": "fhir.com", "term": "term", "label": "label"},{"scheme": "fhir.com1", "term": "term1", "label": "label1"}]'::jsonb);
           insert_resource            
--------------------------------------
 550e8400-e29b-41d4-a716-446655440000
(1 row)

SELECT assert(count(*) = 1, 'insert: not empty observation') FROM observation where logical_id = '550e8400-e29b-41d4-a716-446655440000';
              assert              
----------------------------------
 OK insert: not empty observation
(1 row)

SELECT assert(count(*) = 0, 'insert: empty history') FROM observation_history where logical_id = '550e8400-e29b-41d4-a716-446655440000';
          assert          
--------------------------
 OK insert: empty history
(1 row)

SELECT assert(count(*) = 2, 'insert: not empty tags') FROM observation_tag where resource_id = '550e8400-e29b-41d4-a716-446655440000';
          assert           
---------------------------
 OK insert: not empty tags
(1 row)

SELECT assert(count(*) = 0, 'insert: empty tags history') FROM observation_tag_history where resource_id = '550e8400-e29b-41d4-a716-446655440000';
            assert             
-------------------------------
 OK insert: empty tags history
(1 row)

SELECT update_resource('550e8400-e29b-41d4-a716-446655440000'::uuid,  E'{"resourceType":"Observation","text":{"status":"generated","div":"<div>\\n      <p>\\n        <b>Generated Narrative</b>\\n      </p>\\n      <p>\\n        <b>name</b>: \\n        <span title=\\"Codes: {http://loinc.org 2339-0}\\">Glucose [Mass/volume] in Blood</span>\\n      </p>\\n      <p>\\n        <b>value[x]</b>: 6.3 mmol/l\\n      </p>\\n      <p>\\n        <b>interpretation</b>: \\n        <span title=\\"Codes: {http://hl7.org/fhir/v2/0078 A}\\">abnormal</span>\\n      </p>\\n      <p>\\n        <b>applies[x]</b>: 2-Apr 2013 9:30 --&gt; 5-Apr 2013 9:30\\n      </p>\\n      <p>\\n        <b>issued</b>: 3-Apr 2013 15:30\\n      </p>\\n      <p>\\n        <b>status</b>: final_\\n      </p>\\n      <p>\\n        <b>reliability</b>: ok\\n      </p>\\n      <p>\\n        <b>bodySite</b>: \\n        <span title=\\"Codes: {http://snomed.info/sct 308046002}\\">Superficial forearm vein</span>\\n      </p>\\n      <p>\\n        <b>method</b>: \\n        <span title=\\"Codes: {http://snomed.info/sct 120220003}\\">Injection to forearm</span>\\n      </p>\\n      <p>\\n        <b>identifier</b>: 6323 (official)\\n      </p>\\n      <p>\\n        <b>subject</b>: P. van de Heuvel\\n      </p>\\n      <p>\\n        <b>performer</b>: A. Langeveld\\n      </p>\\n      <h3>ReferenceRanges</h3>\\n      <table class=\\"grid\\">\\n        <tr>\\n          <td>\\n            <b>Low</b>\\n          </td>\\n          <td>\\n            <b>High</b>\\n          </td>\\n          <td>\\n            <b>Meaning</b>\\n          </td>\\n          <td>\\n            <b>Age</b>\\n          </td>\\n        </tr>\\n        <tr>\\n          <td>3.1 mmol/l</td>\\n          <td>6.2 mmol/l</td>\\n          <td> </td>\\n          <td> </td>\\n        </tr>\\n      </table>\\n    </div>"},"name":{"coding":[{"system":"http://loinc.org","code":"2339-0","display":"Glucose [Mass/volume] in Blood"}]},"valueQuantity":{"value":6.3,"units":"mmol/l","system":"http://unitsofmeasure.org","code":"mmol/l"},"interpretation":{"coding":[{"system":"http://hl7.org/fhir/v2/0078","code":"A","display":"abnormal"}]},"appliesPeriod":{"start":"2013-04-02T09:30:10+01:00","end":"2013-04-05T09:30:10+01:00"},"issued":"2013-04-03T15:30:10+01:00","status":"final","reliability":"ok","bodySite":{"coding":[{"system":"http://snomed.info/sct","code":"308046002","display":"Superficial forearm vein"}]},"method":{"coding":[{"system":"http://snomed.info/sct","code":"120220003","display":"Injection to forearm"}]},"identifier":{"use":"official","system":"http://www.bmc.nl/zorgportal/identifiers/observations","value":"6323"},"subject":{"reference":"Patient/f001","display":"P. van de Heuvel"},"performer":[{"reference":"Practitioner/f005","display":"A. Langeveld"}],"referenceRange":[{"low":{"value":3.1,"units":"mmol/l","system":"http://unitsofmeasure.org","code":"mmol/l"},"high":{"value":6.2,"units":"mmol/l","system":"http://unitsofmeasure.org","code":"mmol/l"}}]}'::jsonb, '[{"scheme": "fhir.com", "term": "term", "label": "label"}, {"scheme": "fhir.com2", "term": "term2", "label": "label2"}]'::jsonb);
           update_resource            
--------------------------------------
 550e8400-e29b-41d4-a716-446655440000
(1 row)

SELECT assert(count(*)=1,'update: not empty observation') FROM observation where logical_id = '550e8400-e29b-41d4-a716-446655440000';
              assert              
----------------------------------
 OK update: not empty observation
(1 row)

SELECT assert(count(*)=1,'update: not empty history') FROM observation_history where logical_id = '550e8400-e29b-41d4-a716-446655440000';
            assert            
------------------------------
 OK update: not empty history
(1 row)

SELECT assert(count(*)=3,'update: 3 merged tags') FROM observation_tag where resource_id = '550e8400-e29b-41d4-a716-446655440000';
          assert          
--------------------------
 OK update: 3 merged tags
(1 row)

SELECT assert(count(*)=2,'update: 2 tags in history') FROM observation_tag_history where resource_id = '550e8400-e29b-41d4-a716-446655440000';
            assert            
------------------------------
 OK update: 2 tags in history
(1 row)

SELECT delete_resource('550e8400-e29b-41d4-a716-446655440000'::uuid, 'observation');
           delete_resource            
--------------------------------------
 550e8400-e29b-41d4-a716-446655440000
(1 row)

SELECT assert(count(*)=0,'delete: empty res') FROM observation where logical_id = '550e8400-e29b-41d4-a716-446655440000';
        assert        
----------------------
 OK delete: empty res
(1 row)

SELECT assert(count(*)=2,'delete: 2 in history') FROM observation_history where logical_id = '550e8400-e29b-41d4-a716-446655440000';
         assert          
-------------------------
 OK delete: 2 in history
(1 row)

SELECT assert(count(*)=0,'delete: 0 in tags') FROM observation_tag where resource_id = '550e8400-e29b-41d4-a716-446655440000';
        assert        
----------------------
 OK delete: 0 in tags
(1 row)

SELECT assert(count(*)=5,'delete: 5 in history tags') FROM observation_tag_history where resource_id = '550e8400-e29b-41d4-a716-446655440000';
            assert            
------------------------------
 OK delete: 5 in history tags
(1 row)

ROLLBACK;
ROLLBACK
